### DISCUSSION TITLE
Ensure newline before `typer.secho`

### DISCUSSION THREAD

**Discussion by WilliamDEdwards** [fastapi/typer#661] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the Typer documentation, with the integrated search.
- [X] I already searched in Google "How to X in Typer" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to Typer but to [Click](https://github.com/pallets/click).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
-
```

### Description

How do I tell Typer to ensure a newline before `typer.secho` when there was output before?

In my case, I do not know if there was output before `typer.secho` (as it is called from a generic function). So adding a leading `\n` does not suffice, as that would cause a newline to be present when there was no output before too.

For example:

```
typer.secho('Hi')
typer.secho('Hello')
```

... should output:

```
Hi

Hello
```

... and this:

```
typer.secho('Hello')
```

... should output this:

```
Hello
```

### Operating System

Linux, macOS

### Typer Version

0.7.0

### Python Version

3.11

**Comment by DhruvalAnandkar** [comment]
Hi @WilliamDEdwards! Short version: no, there isn't a built-in flag for this — Typer and Click don't track "has anything been printed yet" for you, so you'll need a small bit of state to handle it yourself.

The core idea: keep a flag that gets set the first time anything is printed, and have your generic function check it before deciding whether to add a separating blank line.

Wrap your regular output calls so they just mark that something's been printed, without adding any spacing themselves:

```python
import typer

_has_output = False

def echo(message: str = "", **kwargs) -> None:
    global _has_output
    typer.secho(message, **kwargs)
    _has_output = True

def echo_section(message: str = "", **kwargs) -> None:
    global _has_output
    if _has_output:
        typer.echo() # blank line separator
    typer.secho(message, **kwargs)
    _has_output = True
```

Note: Swap your regular typer.secho() calls for echo(), and use echo_section() specifically in your generic function. That way normal sequential output stays tight with no extra spacing, but your generic function automatically gets a blank line before it whenever something already printed, and skips it cleanly when it's the very first thing on screen.

One quick heads-up on scaling this up: if the app grows or you start testing commands in isolation, a bare global can get awkward. The same flag works just as well stored on Typer's ctx.obj and passed through your commands instead - happy to sketch that version too if it'd help your setup.

I hope this gets you unblocked! Let me know if this fits your case. (And if this clears things up, I'd really appreciate it if you could mark this as the answer! )

**Comment by svlandeg** [comment]
@DhruvalAnandkar: are all of these answers to discussions AI generated? Please disclose 🙏

**Comment by DhruvalAnandkar** [comment]
Hi @svlandeg , Yes, I used AI to help structure this draft for the explanation because English is not my first language even not second language but I did review the logic and also verified the state-flag approach before posting to make sure it was a solid workaround for @WilliamDEdwards .

I completely understand that I should have explicitly disclosed that upfront that is entirely my mistake I will make sure to clearly label any AI assistance in my future contributions. I still hope the proposed code helps unblock the issue!

**Comment by phalberg** [comment]
Hey @WilliamDEdwards, I don't think this is possible with Typer.
There are however two directions to go, based on your use-case:
- If you need `secho` (which is used for styling mostly), it is advised to use Rich
- If you do not care about styling, then I would use `typer.echo()` but if that is used, then probably just using the normal python `print()` should be better as said in the docs themselves.

So I would refer to those alternatives, as I don't really see an approach in Typer that is supported for this. I do see a previous solution to this in the discussion, but not sure if having a wrapper like that will be very useful for your use-case.

Also this "answer" (not sure if it even is a answer..) is quite late 😓 but it might be useful to anyone that still tries to debug this :)
